aiosmtpd/smtp.py

```diff
@@ -155,48 +155,60 @@ def push(self, msg):
-                arg = line[i+1:].strip()
-            max_sz = (self.command_size_limits[command]
-                      if self.extended_smtp
-                      else self.command_size_limit)
-            if len(line) > max_sz:
-                yield from self.push('500 Error: line too long')
-                continue
-            if (self._tls_handshake_failed
-                    and command != 'QUIT'):                   # pragma: nossl
-                yield from self.push(
-                    '554 Command refused due to lack of security')
-                continue
-            if (self.require_starttls
-                    and (not self._tls_protocol)
-                    and (command not in ['EHLO', 'STARTTLS', 'QUIT'])):
-                # RFC3207 part 4
-                yield from self.push('530 Must issue a STARTTLS command first')
-                continue
-            method = getattr(self, 'smtp_' + command, None)
-            if not method:
-                yield from self.push(
-                    '500 Error: command "%s" not recognized' % command)
-                continue
-            yield from method(arg)
+            try:
+                # XXX this rstrip may not completely preserve old behavior.
+                line = line.decode('utf-8').rstrip('\r\n')
+                log.info('Data: %r', line)
+                if not line:
+                    yield from self.push('500 Error: bad syntax')
+                    continue
+                i = line.find(' ')
+                if i < 0:
+                    command = line.upper()
+                    arg = None
+                else:
+                    command = line[:i].upper()
+                    arg = line[i+1:].strip()
+                max_sz = (self.command_size_limits[command]
+                          if self.extended_smtp
+                          else self.command_size_limit)
+                if len(line) > max_sz:
+                    yield from self.push('500 Error: line too long')
+                    continue
+                if (self._tls_handshake_failed
+                        and command != 'QUIT'):             # pragma: nossl
+                    yield from self.push(
+                        '554 Command refused due to lack of security')
+                    continue
+                if (self.require_starttls
+                        and (not self._tls_protocol)
+                        and (command not in ['EHLO', 'STARTTLS', 'QUIT'])):
+                    # RFC3207 part 4
+                    yield from self.push(
+                        '530 Must issue a STARTTLS command first')
+                    continue
+                method = getattr(self, 'smtp_' + command, None)
+                if not method:
+                    yield from self.push(
+                        '500 Error: command "%s" not recognized' % command)
+                    continue
+                yield from method(arg)
+            except Exception as e:
```

Comment:
Okay, I see what you're saying.  You'd like the server to be as robust as possible, so that if something goes wrong, at least the client will get a 500.

That makes sense.  I think I'd ask just two things then.

* Can you make the local variable `error` instead of `e`?  I dislike abbreviations.
* Can you add something like `log.exception('Client handler error')` to the `except` clause.  I think if the server implementation throws an unexpected exception, it should get logged.